beacon-chain/operations/blstoexecchanges/pool.go

```diff
@@ -0,0 +1,112 @@
+
+import (
+	"sort"
+	"sync"
+
+	"github.com/prysmaticlabs/prysm/v3/beacon-chain/state"
+	fieldparams "github.com/prysmaticlabs/prysm/v3/config/fieldparams"
+	"github.com/prysmaticlabs/prysm/v3/config/params"
+	types "github.com/prysmaticlabs/prysm/v3/consensus-types/primitives"
+	ethpb "github.com/prysmaticlabs/prysm/v3/proto/prysm/v1alpha1"
+)
+
+// PoolManager maintains pending and seen BLS-to-execution-change objects.
+// This pool is used by proposers to insert BLS-to-execution-change objects into new blocks.
+type PoolManager interface {
+	PendingBLSToExecChanges(noLimit bool) []*ethpb.SignedBLSToExecutionChange
+	InsertBLSToExecChange(state state.ReadOnlyBeaconState, change *ethpb.SignedBLSToExecutionChange)
+	MarkIncluded(change *ethpb.SignedBLSToExecutionChange)
+}
+
+// Pool is a concrete implementation of PoolManager.
+type Pool struct {
+	lock    sync.RWMutex
+	pending []*ethpb.SignedBLSToExecutionChange
+}
+
+// NewPool returns an initialized pool.
+func NewPool() *Pool {
+	return &Pool{
+		pending: make([]*ethpb.SignedBLSToExecutionChange, 0),
+	}
+}
+
+// PendingBLSToExecChanges returns objects that are ready for inclusion at the given slot.
+// Without returnAll, this method will not return more than the block enforced MaxBlsToExecutionChanges.
+func (p *Pool) PendingBLSToExecChanges(returnAll bool) []*ethpb.SignedBLSToExecutionChange {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	if returnAll {
+		return p.pending
+	}
+	maxChanges := params.BeaconConfig().MaxBlsToExecutionChanges
+	pending := make([]*ethpb.SignedBLSToExecutionChange, maxChanges)
+	for i, ch := range p.pending {
+		pending[i] = ch
+		if uint64(len(pending)) == maxChanges {
+			break
+		}
+	}
+	return pending
+}
+
+// InsertBLSToExecChange inserts an object into the pool.
+// This method is a no-op if the pending exit already exists or the validator is already exited.
+func (p *Pool) InsertBLSToExecChange(state state.ReadOnlyBeaconState, change *ethpb.SignedBLSToExecutionChange) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	// Prevent malformed messages from being inserted.
+	if change == nil ||
+		change.Message == nil ||
+		len(change.Signature) != fieldparams.BLSSignatureLength ||
+		len(change.Message.FromBlsPubkey) != fieldparams.BLSPubkeyLength ||
+		len(change.Message.ToExecutionAddress) != fieldparams.ExecutionAddressLength {
+		return
+	}
+
+	// Do we already have a pending object for the validator?
+	existsInPending, _ := existsInList(p.pending, change.Message.ValidatorIndex)
+	if existsInPending {
+		return
+	}
+
+	// Does the validator already have ETH1 withdrawal credentials?
+	v, err := state.ValidatorAtIndexReadOnly(change.Message.ValidatorIndex)
+	if err != nil || v.HasETH1WithdrawalCredential() {
+		return
+	}
+
+	// Insert into pending list and sort.
+	p.pending = append(p.pending, change)
+	sort.Slice(p.pending, func(i, j int) bool {
+		return p.pending[i].Message.ValidatorIndex < p.pending[j].Message.ValidatorIndex
+	})
```

Comment:
Newer validators don't need to send this message because they already have an ETH1 withdrawal credential to begin with. I think @potuz will have an answer to how many validators will need to change their credentials. Out of those, I believe preferential treatment of earlier stakers is the right thing to do. Of course we can always randomize it and order by the BLS pubkey.

Reply:
On a related note, we have the same ordering for voluntary exits. Whatever we decide on, I think we should keep these two consistent.

Reply:
> Newer validators don't need to send this message because they already have an ETH1 withdrawal credential
Is this true? They could still send in BLS without knowing ETH1

> On a related note, we have the same ordering for voluntary exits.
I'd argue exit is wrong because it's unfair and we should change that

Reply:
> They could still send in BLS without knowing ETH1

Will need to double check with Potuz, but if this is true then they have ETH1 creds already, and thus any message they send will be ignored because of the `v.HasETH1WithdrawalCredential()` check.

> I'd argue exit is wrong because it's unfair and we should change that

OK, I am convinced that FIFO is better.